### DISCUSSION TITLE
建议matchTime搭配activityIds作用

### DISCUSSION THREAD

**Discussion by jiuqianyuan** [gkd-kit/gkd#376] [Ideas]
当前规则中同时设置了，`matchTime`、`resetMatch:'app',`、`activityIds`，`matchTime`貌似会从进入app就开始匹配倒计时，而不是进入指定activityIds。导致超时后再进入指定activityId时，规则已失效。
比如设置的`支付界面`一次性规则，时间在进入`app首页`就已经花掉了。

建议如规则中指定了`activityIds`，则从进入此规则时，`matchTime`开始生效。

演示用例：
先进入笔记**设置**界面`activityIds:'com.miui.notes.ui.NotesPreferenceActivity'`，等待过了`matchTime`时间后，进入笔记列表界面 `activityIds: 'com.miui.notes.ui.NotesListActivity',`，此时规则已休眠，导致未触发点击。

```
{
  id: 'com.miui.notes',
  name: '笔记',
  groups: [
    {
      key: 0,
      name: '点击测试',
      quickFind: true,
      matchTime: 5000,
      actionMaximum: 1,
      resetMatch: 'app',
      activityIds: 'com.miui.notes.ui.NotesListActivity',
      rules: '@[id="com.miui.notes:id/note_group"] [text^="一"]',
      snapshotUrls: 'https://i.gkd.li/import/13628477',
    },
  ],
}
```
日志
```
************* Log Head ****************
Date of Log        : 2023_12_11
Android SDK        : 33
App VersionName    : 1.5.4
App VersionCode    : 14
************* Log Head ****************

00:11:59.393 D/GkdAbService [DefaultDispatcher-worker-2, li.songe.gkd.service.GkdAbService$1$7$1.emit(GkdAbService.kt:301)]: TopActivity(appId=com.miui.notes, activityId=com.miui.notes.ui.NotesPreferenceActivity)
00:12:14.277 D/GkdAbService [DefaultDispatcher-worker-2, li.songe.gkd.service.GkdAbService$1$7$1.emit(GkdAbService.kt:301)]: args[0] = TopActivity(appId=com.miui.notes, activityId=com.miui.notes.ui.NotesListActivity)
args[1] = subsId:-1, gKey=0, gName:点击测试, ruleIndex:0, rKey:null, active:false

00:12:19.341 D/GkdAbService [DefaultDispatcher-worker-6, li.songe.gkd.service.GkdAbService$1$7$1.emit(GkdAbService.kt:301)]: TopActivity(appId=li.songe.gkd.debug, activityId=li.songe.gkd.MainActivity)

```

**Comment by lisonge** [comment]
没看懂你想表达什么？你说的这些不都是正常的预期行为吗？

**Reply by jiuqianyuan** [reply]
实际上并没有，如日志所示，我进入指定activityId=com.miui.notes.ui.NotesListActivity的界面后，规则已经失效
`args[1] = subsId:-1, gKey=0, gName:点击测试, ruleIndex:0, rKey:null, active:false`

**Comment by jiuqianyuan** [comment]
https://github.com/gkd-kit/gkd/blob/7ad5917f710f82428fdf4f0934f2ccaeb83470b2/app/src/main/kotlin/li/songe/gkd/service/AbState.kt#L128-L139
此处是否应改为如下，否则`resetMatch: 'app',` ，进入指定activity中时，可能已经`(t - appChangeTime > rule.matchAllTime)`


    if (rule.matchTime != null) {
        // 超出了匹配时间
        if (t - activityChangeTime > rule.matchAllTime) {
            return false
        }
    }

**Comment by jiuqianyuan** [comment]
不再调整matchTime逻辑，改为调整使用方法
- https://github.com/gkd-kit/subscription/pull/3120

**Comment by wuwawo** [comment]
key: 1,
      name: '版本更新',
      quickFind: true,
      matchLauncher: true,
      matchTime: 10000,
      actionMaximum: 1,
      resetMatch: 'app',
      activityIds: [
        'com.cmbchina.ccd.pluto.cmbActivity.CMBRootActivityV2',
        'com.cmbchina.ccd.pluto.cmbActivity.secplugin',
      ],
      rules:
        '[id="com.cmbchina.ccd.pluto.cmbActivity:id/bt_sure_tips"] - [id="com.cmbchina.ccd.pluto.cmbActivity:id/bt_cancel_tips"]',
      snapshotUrls: [
        'https://i.gkd.li/import/12647025',
        'https://i.gkd.li/import/12727203',
        'https://i.gkd.li/import/13345771', // matchLauncher: true,
      ],
`掌上生活`  中  `版本更新`  规则 ，我也遇到过类似的问题，规则遇到触发`activityIds`时可能发生在触发`resetMatch: 'app'`10s后，需要将APP切到后台，再次进入才能跳过

这种指定了activitylds的情况是不是可以使用`resetMatch:'activityld'`或者直接删掉`resetMatch:'app'`,毕竟`resetMatch:'app'`主要方便适配无指定activitylds的matchTime时间内开屏类广告。

还有一个问题：`matchLauncher: true`,表达的什么意思?
